content/articles/2018-02-rlang-0.2.0.Rmd

```diff
@@ -0,0 +1,228 @@
+date: 2018-02-26
+categories: [package]
+photo:
+  url: https://unsplash.com/photos/K5gBXHSvtkc
+  author: Denise Johnson
+---
+
+```{r setup, include = FALSE}
+library("rlang")
+knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
+```
+
+```{r crayon, include = FALSE}
+colourise_chunk <- function(type) {
+  function(x, options) {
+    x <- pkgdown:::escape_html(x)
+    x <- ansistrings::ansi_to_html(x, fullpage = FALSE)
+    class <- tolower(options$engine)
+    sprintf('<pre class="knitr %s">%s</pre>\n', class, x)
+  }
+}
+
+knitr::knit_hooks$set(
+  output = colourise_chunk("output"),
+  message = colourise_chunk("message"),
+  warning = colourise_chunk("warning"),
+  error = colourise_chunk("error")
+)
+
+options(crayon.enabled = FALSE)
+```
+
+We are happy to announce a new version of [rlang](http://rlang.tidyverse.org/)
+that features many improvements to tidy evaluation:
+
+* The quirks in the quasiquotation syntax have been much reduced.
+
+* Quosures gained a much improved printing method with colour support.
+
+* The performance of quoting, splicing and quosure evaluation was
+  vastly improved.
+
+* Many bugs have been fixed.
+
+
+Install the latest version of rlang with:
+
+```{r, eval = FALSE}
+install.packages("rlang")
+```
+
+
+### Syntax adjustments for the !! operator
+
+The headline improvement is a much improved syntax for the `!!`
+operator (pronounced bang-bang). In a way bang-bang is a lie, or
+a [polite fiction](https://adv-r.hadley.nz/quasiquotation.html#the-polite-fiction-of).
+It is not a real R operator and only works in tidy evaluation
+functions. To implement `!!`, we simply look at the code and
+reinterpret two consecutive `!` negations as the unquoting
+operator. The major downside of this approach was that our `!!`
+operator inherited from the operation precedence of `!`, which is very
+low. As a result `!!a + b` was interpreted by R as `!!(a + b)` rather
+than `(!!a) + b`. To work around this, tidy eval users had to add
+explicit parentheses, which was often confusing.
+
+Happily this is no longer necessary. We now reprocess the whole syntax
+tree to give `!!` a more natural operation precedence, i.e. that of
+unary `+` and `-`. This means that expressions like `!!a > b` now do
+the expected thing (only `a` is unquoted):
+
+```{r}
+a <- sym("foo")
+expr(!!a > b)
+```
+
+In addition, we automatically strip one layer when `!!` is wrapped in
+parentheses. This is useful for unquoting function names cleanly:
+
+```{r}
+expr((!!a)(bar, baz))
+
+op <- sym(">")
+expr((!!op)(bar, baz))
+```
+
+
+### A new print method for quosures
+
+#### Problems with base R deparsing
+
+When R code is printed at the console (be it a function, an
+expression, a formula or a quosure), R runs the function `deparse()`
+to transform the code to a printable string. The deparsing mechanism
+in base R is not optimal for tidy eval:
+
+  - It adds redundant parentheses between the `!` of `!!`:
+
+    ```{r}
+    quote(!!x)
```

Comment:
The prose gets us prepared to see `deparse()` and yet the code shows `quote()`, with an implicit deparse. Can you make this more explicit? Could this be `deparse(expr(!!x))`?